File: code/backend/app/models/assistant.py
"""
Step 18 AI 抽屉的请求/响应 Pydantic 模型。

跟其他 step 的 models 一样, 用 Pydantic v2 做请求体校验 + 文档自动生成。
响应结构走统一的 {code, message, data} envelope (response.success/error 包装),
这里只列请求体 + 必要的响应 data 子结构。
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class CreateSessionRequest(BaseModel):
    title: str | None = Field(None, description="会话标题, 不填默认按首条问题自动生成")
    context: dict[str, Any] | None = Field(
        None,
        description="会话上下文: {site_id, building_id, time_range, metric}",
    )


class SendMessageRequest(BaseModel):
    """发消息请求体。

    content: 用户输入的文本
    context: 当前上下文覆盖 (前端从 contextStore 拿, 后端拼进 system prompt)
    """
    content: str = Field(..., min_length=1, max_length=4000, description="用户消息内容")
    context: dict[str, Any] | None = Field(
        None,
        description="当前上下文覆盖, 不传则用 session.context",
    )


class SessionResponse(BaseModel):
    id: str
    title: str = ""
    context: dict[str, Any] = {}
    created_at: str
    updated_at: str | None = None
    message_count: int = 0


class MessageResponse(BaseModel):
    id: str
    session_id: str | None = None
    role: str
    content: str
    tool_calls: list[Any] = []
    citations: list[Any] = []
    referenced_chunk_ids: list[str] = []
    referenced_point_ids: list[str] = []
    optimization_plan: dict[str, Any] | None = None
    tokens_used: int | None = None
    latency_ms: int | None = None
    created_at: str


class OptimizationPlanResponse(BaseModel):
    """节能优化结构化数据响应。"""
    message_id: str
    has_plan: bool
    plan: dict[str, Any] | None = None

File: code/frontend/src/api/assistant.ts
// ============================================================================
// Assistant API - AI 抽屉的 REST 接口包装
// ----------------------------------------------------------------------------
// 6 个 REST 接口 (SSE 流式 sendMessage 不在这里, 在 stores/ai.ts 里手写
// fetch + ReadableStream, 因为 EventSource 不支持 POST + JWT header)。
//
// 后端响应统一走 {code, message, data} envelope, api.get/post 已经自动 unwrap
// data 字段, 调用方拿到的是 T。
// ============================================================================

import { api } from './client'

// ---------------------------------------------------------------------------
// 类型定义
// ---------------------------------------------------------------------------

export interface ChatContext {
  site_id?: string
  site_name?: string
  building_id?: string
  building_name?: string
  time_range?: { start: string; end: string }
  metric?: string
  scenario?: 'optimization' | 'anomaly' | 'standard' | 'data'
}

export interface ChatSession {
  id: string
  title: string
  context: Record<string, unknown>
  created_at: string
  updated_at?: string
  message_count: number
}

export interface ToolCall {
  id: string
  name: string
  arguments: Record<string, unknown>
  ok?: boolean
}

export interface Citation {
  type: 'data' | 'anomaly' | 'document'
  [key: string]: unknown
}

export interface OptimizationPlan {
  summary: string
  problems: Array<{
    problem_desc: string
    evidence_ref: string
    severity: 'high' | 'medium' | 'low'
  }>
  measures: Array<{
    measure_name: string
    saved_kwh: number
    saved_co2: number
    payback_months: number
    difficulty: 'easy' | 'medium' | 'hard'
    clause_ref: string
  }>
  priorities: Array<{
    measure_name: string
    reason: string
  }>
}

export interface ChatMessage {
  id: string
  session_id?: string
  role: 'user' | 'assistant' | 'tool' | 'system'
  content: string
  tool_calls?: ToolCall[]
  citations?: Citation[]
  referenced_chunk_ids?: string[]
  referenced_point_ids?: string[]
  optimization_plan?: OptimizationPlan | null
  tokens_used?: number
  latency_ms?: number
  created_at: string
}

// ---------------------------------------------------------------------------
// 会话管理
// ---------------------------------------------------------------------------

export async function createSession(opts?: {
  title?: string
  context?: ChatContext
}): Promise<ChatSession> {
  return api.post<ChatSession>('/assistant/sessions', {
    title: opts?.title,
    context: opts?.context,
  })
}

export async function listSessions(limit = 50): Promise<{ items: ChatSession[]; total: number }> {
  return api.get('/assistant/sessions', { params: { limit } })
}

export async function listMessages(sessionId: string, limit = 100): Promise<{ items: ChatMessage[]; total: number }> {
  return api.get(`/assistant/sessions/${sessionId}/messages`, { params: { limit } })
}

// ---------------------------------------------------------------------------
// 节能优化 + PDF
// ---------------------------------------------------------------------------

export async function exportPdf(messageId: string): Promise<Blob> {
  // 用 client 直接调, 因为要拿 raw blob 不能让 api 拦截器 unwrap
  const client = (await import('./client')).default
  const resp = await client.post(`/assistant/messages/${messageId}/export-pdf`, null, {
    responseType: 'blob',
  })
  return resp.data as Blob
}
